Imageassembler - Version 0.1.0

Nicholas-Schaub opened the pull request:
This PR adds a plugin that will stitch images together using an input stitching vector. NIST has an image assembling plugin, which is highly optimized. However, it only supports images up to 2^32 pixels in size. To stitch images smaller than 2^32 pixels, the NIST image assembling plugin should be used. To stitch images larger than 2^32 pixels, this plugin should be used.

In the future, this plugin should be rewritten in Java. The interface between Python becomes exponentially slower as the size of the image increases. Rewriting in Java should drastically increase performance.

Nicholas-Schaub commented:
@NHotaling , @sunnielyu Waiting for review.

NHotaling commented on polus-image-assembler-plugin/README.md:
````diff
@@ -0,0 +1,37 @@
+x002_y002.ome.tif
+```
+
+Then the output image name will be `x00<0-1>_y00<0-1>.ome.tif`. However, if the images range from `000` to `010` trailing the x and y positions of the image names, then the output image name will be `x0<00-19>_y00<00-19>.ome.tif` because the 2nd digit ranges from 0-1 and the 3rd digit ranges from 0-9. However, this could give the impression that the resulting image contains a 20x20 grid of images ...
+
````
We should probably just flag this as an issue as well as being here in the comments

NHotaling commented on polus-image-assembler-plugin/src/imagesize.py:
```diff
@@ -0,0 +1,319 @@
+import re
+import struct
+from xml.etree import ElementTree
+
+_UNIT_KM = -3
```
Remove and call from bfio

NHotaling commented on polus-image-assembler-plugin/src/main.py:
```diff
@@ -0,0 +1,296 @@
+    try:
+        vectors = [str(p.absolute()) for p in Path(stitchPath).iterdir() if p.is_file() and "".join(p.suffixes)=='.txt']
+    except FileNotFoundError:
+        # Workaround for WIPP bug
+        if Path(stitchPath).name == 'metadata_files':
+            stitchPath = str(Path(imgPath).joinpath('metadata_files').absolute())
+            imgPath = str(Path(imgPath).joinpath('images').absolute())
+            vectors = [str(p.absolute()) for p in Path(stitchPath).iterdir() if p.is_file() and "".join(p.suffixes)=='.txt']
```
If this is a reported bug in their repo maybe link to that bug so its more clear why this exists. Just reading it, it looks like there are files that you want to not "see" but it'd be good and easier to just link to the github issue if possible.

Nicholas-Schaub replied:
At the time I wrote this, it was considered a bug. It is now just standard practice to always detect what input folder you are being fed and switch to an `images` subdirectory if it's present.

Nicholas-Schaub commented:
Vastly simplified code (removed 3 files).
Added the new bfio with Python backend.
Removed subprocessing module since new bfio backend does not have issues slowing down as the number of images opened increases.
Tests show this version runs in 2 hours on 55 images roughly 2GB in size, compared to 24 hours using the Java backend.
